crates/telio-batcher/src/batcher.rs

```diff
@@ -0,0 +1,533 @@
+use async_trait::async_trait;
+use futures::{stream::FuturesUnordered, Future, StreamExt};
+use std::collections::HashMap;
+use telio_task::{task_exec, BoxAction, ExecError, Runtime, Task};
+use telio_utils::{telio_log_debug, telio_log_warn};
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Generic error")]
+    Generic(#[from] ExecError),
+}
+
+#[cfg_attr(any(test, feature = "mockall"), mockall::automock)]
+#[async_trait]
+pub trait BatcherTrait {
+    async fn remove(&self, key: String);
+    async fn add(
+        &self,
+        key: String,
+        interval: std::time::Duration,
+        threshold: std::time::Duration,
+    ) -> std::result::Result<tokio::sync::watch::Receiver<()>, Error>;
+}
+
+struct State {
+    notifiers: HashMap<String, BatchEntry>,
+}
+
+pub struct Batcher {
+    task: Task<State>,
+}
+
+struct BatchEntry {
+    deadline: tokio::time::Instant,
+    interval: std::time::Duration,
+    threshold: std::time::Duration,
+    tx: tokio::sync::watch::Sender<()>,
```

Comment:
Could you elaborate a bit more how those could be missed? There's supposed to be 1:1 sender:receiver mapping by the current design

Reply:
Under heavy load and short interval, the `select!` in the SessionKeepers' Task might not be called often enough to notice all notifications. But thinking about it now, this might actually be a desired behaviour... so I think you can ignore this comment.